lib/web_ui/flutter_js/src/types.d.ts

```diff
@@ -0,0 +1,73 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+type JSCompileTarget = "dart2js" | "dartdevc";
+type WasmCompileTarget = "dart2wasm";
+
+export type CompileTarget = JSCompileTarget | WasmCompileTarget;
+
+export type WebRenderer =
+  "html" |
+  "canvaskit" |
+  "skwasm";
+
+interface ApplicationBuildBase {
+  renderer: WebRenderer;
+}
+
+export interface JSApplicationBuild extends ApplicationBuildBase {
+  compileTarget: JSCompileTarget;
+  mainJsPath: string;
+}
+
+export interface WasmApplicationBuild extends ApplicationBuildBase {
+  compileTarget: WasmCompileTarget;
+  mainWasmPath: string;
+  jsSupportRuntimePath: string;
+}
+
+export type ApplicationBuild = JSApplicationBuild | WasmApplicationBuild;
```

Comment:
Do you need this union? Wouldn't it be equivalent that `interface ApplicationBuildBase` is called `interface ApplicationBuild`?

Reply:
A union like this is super useful in TS. An `interface` basically has open polymorphism, but a union has closed polymorphism. This basically means that if you're using the union (`ApplicationBuild`), you can actually differentiate based on the `compileTarget` field, and the compiler will actually refine the type in the the subsequent branches. This won't happen when using the interface (and in fact, the `compileTarget` field doesn't exist on the interface, so you couldn't check for it). Example:
```
const build: ApplicationBuild = //something;
// build is an ApplicationBuild type
if (build.compileTarget === "dart2wasm") {
  // build is actually a WasmApplicationBuild in this branch, so I can access fields specific to WasmApplicationBuild
  console.log(build.mainWasmPath);
} else {
  // the compiler can even tell that the only remaining possible values are of type JSApplicationBuild
  console.log(build.mainJsPath);
}
```